File: src/main/kotlin/com/eunmin/arrow/repository/UserRepository.kt
package com.eunmin.arrow.repository

import arrow.core.Either
import com.eunmin.arrow.domain.UserException
import com.eunmin.arrow.dto.UserDocument

interface UserRepository {
    fun create(userDocument: UserDocument): Either<UserException, UserDocument>
    fun get(id: String): Either<UserException, UserDocument>
}

File: scala-with-cats/src/main/scala/com/eunmin/cats/repository/UserRepository.scala
package com.eunmin.cats.repository

import cats.effect.IO
import com.eunmin.cats.domain.UserException
import com.eunmin.cats.dto.UserDocument

trait UserRepository {
  def create(userDocument: UserDocument): IO[Either[UserException, UserDocument]]
  def get(id: String): IO[Either[UserException, UserDocument]]
}
